lib/cog/command/pipeline/initializer.ex

```diff
@@ -68,4 +83,120 @@ defmodule Cog.Command.Pipeline.Initializer do
     end
   end
 
+  defp self_register_user(request, true, state) do
+    sender = request["sender"]
+    case Users.by_chat_handle(sender["handle"], request["adapter"]) do
+      {:ok, _} ->
+        :ok
+      {:error, :not_found} ->
+        case find_available_username(sender["handle"]) do
+          {:ok, username} ->
+            case Users.new(%{"username" => username,
+                             "first_name" => sender["first_name"],
+                             "last_name" => sender["last_name"],
+                             "email_address" => sender["email"],
+                             "password" => Passwords.generate_password(12)}) do
+              {:ok, user} ->
+                case ChatHandles.set_handle(user, request["adapter"], sender["handle"]) do
+                  {:ok, _} ->
+                    alert_self_registration_success(user, request, state)
+                    :ok
+                  error ->
+                    Logger.error("Failed to auto-register user '#{sender["handle"]}': #{inspect error}")
+                    alert_self_registration_failed(request, state)
+                    error
+                end
+              error ->
+                Logger.error("Failed to auto-register user '#{sender["handle"]}': #{inspect error}")
+                alert_self_registration_failed(request, state)
+                :error
+            end
+          _ ->
+            Logger.error("Failed to auto-register user '#{sender["handle"]}'. No suitable username was found.")
+            alert_self_registration_failed(request, state)
+            :error
+        end
+    end
+  end
+  defp self_register_user(_, _, _) do
+    :ok
+  end
+
+  defp alert_self_registration_success(user, request, state) do
+    {:ok, message} = self_registration_success_message(user, request)
+    publish_response(message, request["room"], request["adapter"], state)
+  end
+  defp alert_self_registration_failed(request, state) do
+    {:ok, message} = self_registration_failed_message(request)
+    publish_response(message, request["room"], request["adapter"], state)
+  end
+
+  defp publish_response(message, room, adapter, state) do
```

Comment:
The code comments in the body of this function that were copied from the original implementation. For instance, *this* particular function should never be called from the context of the `http` adapter, but it's written and commented as though it could.